lib/pluginmanager/command.rb

```diff
@@ -47,6 +47,31 @@ def remove_unused_locally_installed_gems!
     end
   end
 
+  def remove_orphan_dependencies!
+    locked_gem_names = ::Bundler::LockfileParser.new(File.read(LogStash::Environment::LOCKFILE)).specs.map(&:full_name).to_set
+    orphan_gem_specs = ::Gem::Specification.each
+                                           .reject(&:stubbed?) # skipped stubbed (uninstalled) gems
+                                           .reject(&:default_gem?) # don't touch jruby-included default gems
+                                           .reject{ |spec| locked_gem_names.include?(spec.full_name) }
+                                           .sort
+
+    inactive_plugins, orphaned_dependencies = orphan_gem_specs.partition { |spec| LogStash::PluginManager.logstash_plugin_gem_spec?(spec) }
+
+    # uninstall plugins first, to limit damage should one fail to uninstall
+    inactive_plugins.each { |plugin| uninstall_gem!("inactive plugin", plugin) }
+    orphaned_dependencies.each { |dep| uninstall_gem!("orphaned dependency", dep) }
+  end
+
+  def uninstall_gem!(desc, spec)
+    require "rubygems/uninstaller"
```

Comment:
Is this expensive to load on startup? Why do we wait until method is called to load implementation?

Reply:
I think there are a couple reasons why I did it this way.

First, it follows the prior art in this source file.

But I think the bigger factor is that the plugin manager is invoked stand-alone, and is not typically a long-running process, so there's less of a stability benefit to eager-loading. I don't _believe_ that having the uninstaller present will have side-effects for the non-mutation `list` command, but I'd rather not take chances.